packages/dd-trace/src/appsec/rule_manager.js

```diff
@@ -26,137 +20,66 @@ function loadRules (config) {
-  const batch = new Set()
-
-  const newRulesData = new SpyMap(appliedRulesData)
-  let newRuleset
-  let newRulesetId
-  const newRulesOverride = new SpyMap(appliedRulesOverride)
-  const newExclusions = new SpyMap(appliedExclusions)
-  const newCustomRules = new SpyMap(appliedCustomRules)
   const newActions = new SpyMap(appliedActions)
 
+  let wafUpdated = false
+  let wafUpdatedSuccess = true
+
   for (const item of toUnapply) {
-    const { product, id } = item
+    if (!['ASM_DD', 'ASM_DATA', 'ASM'].includes(item.product)) continue
 
-    if (product === 'ASM_DATA') {
-      newRulesData.delete(id)
-    } else if (product === 'ASM_DD') {
-      if (appliedRulesetId === id) {
-        newRuleset = defaultRules
+    try {
+      waf.wafManager.remove(item.path)
+      item.apply_state = ACKNOWLEDGED
+      wafUpdated = true
+
+      if (item.product === 'ASM') {
+        newActions.delete(item.id)
       }
-    } else if (product === 'ASM') {
-      newRulesOverride.delete(id)
-      newExclusions.delete(id)
-      newCustomRules.delete(id)
-      newActions.delete(id)
+    } catch (e) {
+      wafUpdatedSuccess = false
+      item.apply_state = ERROR
+      item.apply_error = e.toString()
+      Reporter.reportWafConfigError(waf.wafManager.ddwafVersion, waf.wafManager.rulesVersion)
     }
   }
 
   for (const item of [...toApply, ...toModify]) {
-    const { product, id, file } = item
+    if (!['ASM_DD', 'ASM_DATA', 'ASM'].includes(item.product)) continue
 
-    if (product === 'ASM_DATA') {
-      if (file && file.rules_data && file.rules_data.length) {
-        newRulesData.set(id, file.rules_data)
-      }
+    try {
+      const updateResult = waf.wafManager.update(item.product, item.file, item.path)
+      item.apply_state = updateResult.success ? ACKNOWLEDGED : ERROR
 
-      batch.add(item)
-    } else if (product === 'ASM_DD') {
-      if (appliedRulesetId && appliedRulesetId !== id && newRuleset !== defaultRules) {
-        item.apply_state = ERROR
-        item.apply_error = 'Multiple ruleset received in ASM_DD'
+      if (updateResult.success) {
+        wafUpdated = true
+        Reporter.reportSuccessfulWafUpdate(item.product, item.id, updateResult.diagnostics)
       } else {
-        if (file?.rules?.length) {
-          const { version, metadata, rules, processors, scanners } = file
-
-          newRuleset = { version, metadata, rules, processors, scanners }
-          newRulesetId = id
-        }
-
-        batch.add(item)
-      }
-    } else if (product === 'ASM') {
-      if (file?.rules_override?.length) {
-        newRulesOverride.set(id, file.rules_override)
-      }
-
-      if (file?.exclusions?.length) {
-        newExclusions.set(id, file.exclusions)
-      }
-
-      if (file?.custom_rules?.length) {
-        newCustomRules.set(id, file.custom_rules)
+        wafUpdatedSuccess = false
+        item.apply_error = JSON.stringify(extractErrors(updateResult.diagnostics))
```

Comment:
The errors are handled very differently depending on where it happens while the overall logical seems to almost be identical. Should they not be handled all the same? :)